### ISSUE TITLE
Issue managing downloads from multiple tabs when using a remote allocator

### ISSUE THREAD

**Issue by onsi** [chromedp/chromedp#1267]
Hey all - thanks for this great project.  I'm building on top of it to provide a [Ginkgo](https://github.com/onsi/ginkgo) and [Gomega](https://github.com/onsi/gomega) friendly web testing framework.  My goal is to enable parallel testing by spinning up multiple remote allocators and connecting to a single browser instance.  So far everything is working well but I've run into an issue with managing downloads.

#### What versions are you running?

```console
$ go list -m github.com/chromedp/chromedp
github.com/chromedp/chromedp v0.8.7
$ /Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --version
Google Chrome 110.0.5481.100
$ go version
go version go1.19.5 darwin/arm64
```

#### What did you do? Include clear steps.

It's straightforward but verbose.  I have a complete reproducer here:

https://github.com/onsi/download-reproducer

that you can clone and run the output of.  Just `go run main.go`.

#### What did you expect to see?

I expected all four downloads to complete successfully

#### What did you see instead?

I see that the fourth download is cancelled.

More to the point, I'm doing the following:

- create an ExecAllocator
- spin up the browser and grab its websocket url
- create two remote allocators connecting to the websocket URL.  A and B
- create a tab on allocator A (call it `tabA`)
- configure `tabA` to receive downloads with events enabled and `browser.SetDownloadBehaviorBehaviorAllowAndName`
- configure `tabA` to handle `browser.EventDownloadWillBegin` and `browser.EventDownloadProgress`
- use `tabA` to navigate to a site, click on a download button, and download a file
- use `tabA` to navigate to a site, click on a download button, and download a file (again - to prove we can do it twice)
- create a tab on allocator  B (call it `tabB`)
- configure `tabB` with the same `browser.SetDownloadBehaviorBehaviorAllowAndName` (I find this is necessary in order to be able to receive download events on `tabB`
- use **`tabA`** to navigate to a site, click on a download button, and download a file (again - to prove it works even after `tabB` is up and running and has set its own download behavior configuration)
- **close `tabB`**
- use **`tabA`** to navigate to a site, click on a download button, and download a file (one last time - a fourth download to see if we can still download files)
- observe that the fourth download is cancelled

This is preventing me from successfully managing multiple tabs with downloads in parallel.  But more than my usecase this may be a bug in chromedp or chrome's implementation of the protocol (?)

I'm happy to help do whatever's necessary to debug this - I've looked at the chromedp source and no obvious issue has jumped out.  The repo I'm providing turns on debug logging so you can see the events flowing.

Some observations:

- this issue does **not** occur if I create `tabA` and `tabB` using the `ExecAllocator`:

```
// this is fine:
tabA, _ := chromedp.NewContext(execAllocCtx)
tabB, _ := chromedp.NewContext(execAllocCtx)
```

- this issue **does** occur if I create `tabB` with `tabA` as a parent:

```
// this is not fine - closing tabB after configuring it's `SetDownloadBehavior` will cause `tabA`'s downloads to fail:
tabA, _ := chromedp.NewContext(execAllocCtx)
tabB, _ := chromedp.NewContext(tabB)
```

- this issue does not occur if I do not set download behavior on `tabB` (i.e. it's not enough to just open and close a random tab)

Any help would be much appreciated!

**Comment by ZekeLu** [comment]
Thank you @onsi ! This is an excellent issue report!!! 👍 👍 👍 

I haven't took a deep look into the issue. But  I know that there is a workaround.

First of all, it's tricky to call `browser.SetDownloadBehavior` several times from different tabs. There is a reason to move  it from the `page` domain to the `browser` domain (https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-setDownloadBehavior is marked as deprecated).

Here are two comments from https://bugs.chromium.org/p/chromium/issues/detail?id=1204880:

> ... and the download instrumentation is generally target-agnostic ...

>  the Browser domain download support wasn't originally tested with a frame target (the expected use cases implied it being used on the Browser target)

The workaround is to limit `browser.SetDownloadBehavior` to a browser context (please not the **context** here has nothing to do with the Go `context` package).

```diff
@@ -113,10 +115,24 @@ func main() {
    fmt.Print("+++ CREATING NEW TAB\n\n")
    tabB, cancelB := chromedp.NewContext(allocCtxB, chromedp.WithDebugf(log.Printf))

+  Create a browser context.
+   var browserContextID cdp.BrowserContextID
+   if err := chromedp.Run(tabB, chromedp.ActionFunc(func(ctx context.Context) error {
+       c := chromedp.FromContext(ctx)
+       var err error
+       browserContextID, err = target.CreateBrowserContext().
+           WithDisposeOnDetach(true).
+           Do(cdp.WithExecutor(ctx, c.Browser))
+
+       return err
+   })); err != nil {
+       panic(err)
+   }
+
    //and we configure it for downloads
    fmt.Print("+++ CONFIGURING NEW TAB DOWNLOAD BEHAVIOR\n\n")
    chromedp.Run(tabB, browser.SetDownloadBehavior(browser.SetDownloadBehaviorBehaviorAllowAndName).
        WithDownloadPath("./").
+       WithBrowserContextID(browserContextID).
        WithEventsEnabled(true))

```

<details><summary>Here is the full demo:</summary>

```go
package main

import (
	"context"
	"fmt"
	"log"
	"net/http"
	"os"

	"github.com/chromedp/cdproto/browser"
	"github.com/chromedp/cdproto/cdp"
	"github.com/chromedp/cdproto/target"
	"github.com/chromedp/chromedp"
)

const DOWNLOAD_PAGE = `<!DOCTYPE html>
<html lang="en-US">
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width" />
    <title>Downloads Testpage</title>
</head>
<body>
    <button id="download">Download</button>
    <script>
        document.getElementById("download").addEventListener("click", () => {
            let anchor = document.createElement('a');
            anchor.setAttribute('href', 'data:text/plain;charset=utf-8,CONTENT');
            anchor.setAttribute('download', 'file.txt');
            anchor.click()
        })
    </script>
</body>
</html>`

func page(w http.ResponseWriter, req *http.Request) {
	fmt.Fprintf(w, DOWNLOAD_PAGE)
}

func startServer() {
	http.HandleFunc("/page", page)
	go http.ListenAndServe(":8080", nil)
}

func main() {
	startServer()

	// first we start up a browser instance and grab its websocket url
	fmt.Print("+++ SPINNING UP THE BROWSER\n\n")
	opts := append(chromedp.DefaultExecAllocatorOptions[:],
		chromedp.Flag("remote-debugging-port", "9222"),
		// chromedp.Flag("enable-logging", true),
		// chromedp.Flag("v", "100"),
		chromedp.CombinedOutput(log.Writer()),
	)
	execAllocCtx, cancelAllocCtx := chromedp.NewExecAllocator(context.Background(), opts...)
	defer cancelAllocCtx()
	startBrowserCtx, cancelStartBrowserCtx := chromedp.NewContext(execAllocCtx)
	defer cancelStartBrowserCtx()
	if err := chromedp.Run(startBrowserCtx); err != nil {
		panic(err)
	}

	wsURL := "ws://127.0.0.1:9222"

	//now we attach to the browser twice
	allocCtxA, canceldAlloCtxA := chromedp.NewRemoteAllocator(context.Background(), wsURL)
	defer canceldAlloCtxA()
	allocCtxB, canceldAlloCtxB := chromedp.NewRemoteAllocator(context.Background(), wsURL)
	defer canceldAlloCtxB()

	//we create a new tab on A
	fmt.Print("+++ CREATING FIRST TAB\n\n")
	tabA, cancelA := chromedp.NewContext(allocCtxA, chromedp.WithDebugf(log.Printf))
	defer cancelA()

	//and configure it for downloads
	fmt.Print("+++ CONFIGURING FIRST TAB DOWNLOAD BEHAVIOR\n\n")
	chromedp.Run(tabA, browser.SetDownloadBehavior(browser.SetDownloadBehaviorBehaviorAllowAndName).
		WithDownloadPath("./").
		WithEventsEnabled(true))

	downloads := make(chan string, 1)
	chromedp.ListenTarget(tabA, func(ev interface{}) {
		switch ev := ev.(type) {
		case *browser.EventDownloadWillBegin:
			fmt.Print("\n+++ DOWNLOAD BEGIN\n\n")
		case *browser.EventDownloadProgress:
			switch ev.State {
			case browser.DownloadProgressStateCanceled:
				fmt.Print("\n+++ DOWNLOAD CANCELED\n\n")
				downloads <- "CANCELED"
			case browser.DownloadProgressStateCompleted:
				fmt.Print("\n+++ DOWNLOAD COMPLETE\n\n")
				content, _ := os.ReadFile("./" + ev.GUID)
				downloads <- "COMPLETE: " + string(content)
			}
		}
	})

	//and we try downloading something
	fmt.Print("\n+++ LOADING PAGE AND PERFORMING FIRST DOWNLOAD\n\n")
	chromedp.Run(tabA,
		chromedp.Navigate(`http://localhost:8080/page`),
		chromedp.Click(`#download`, chromedp.ByQuery),
	)
	firstDownload := <-downloads

	//and we do it again, to prove that we can
	fmt.Print("+++ PERFORMING SECOND DOWNLOAD\n\n")
	chromedp.Run(tabA, chromedp.Click(`#download`, chromedp.ByQuery))
	secondDownload := <-downloads

	//now we create a new tab in the separate remote allocator
	fmt.Print("+++ CREATING NEW TAB\n\n")
	tabB, cancelB := chromedp.NewContext(allocCtxB, chromedp.WithDebugf(log.Printf))

	var browserContextID cdp.BrowserContextID
	if err := chromedp.Run(tabB, chromedp.ActionFunc(func(ctx context.Context) error {
		c := chromedp.FromContext(ctx)
		var err error
		browserContextID, err = target.CreateBrowserContext().
			WithDisposeOnDetach(true).
			Do(cdp.WithExecutor(ctx, c.Browser))

		return err
	})); err != nil {
		panic(err)
	}

	//and we configure it for downloads
	fmt.Print("+++ CONFIGURING NEW TAB DOWNLOAD BEHAVIOR\n\n")
	chromedp.Run(tabB, browser.SetDownloadBehavior(browser.SetDownloadBehaviorBehaviorAllowAndName).
		WithDownloadPath("./").
		WithBrowserContextID(browserContextID).
		WithEventsEnabled(true))

	// back on tabA, we try another download
	fmt.Print("\n+++ PERFORMING THIRD DOWNLOAD (ON ORIGINAL TAB)\n\n")
	chromedp.Run(tabA, chromedp.Click(`#download`, chromedp.NodeVisible))
	thirdDownload := <-downloads

	// now we close tab B
	fmt.Print("+++ CLOSING NEW TAB\n\n")
	cancelB()

	// and try one last download on tab A
	fmt.Print("\n+++ PERFORMING FOURTH DOWNLOAD (ON ORIGINAL TAB)\n\n")
	chromedp.Run(tabA, chromedp.Click(`#download`, chromedp.NodeVisible))
	fourthDownload := <-downloads

	fmt.Printf("+++ DOWNLOAD #1: EXPECTED 'COMPLETE: CONTENT', GOT '%s'\n", firstDownload)  // this is fine
	fmt.Printf("+++ DOWNLOAD #2: EXPECTED 'COMPLETE: CONTENT', GOT '%s'\n", secondDownload) // this is fine
	fmt.Printf("+++ DOWNLOAD #3: EXPECTED 'COMPLETE: CONTENT', GOT '%s'\n", thirdDownload)  // this is fine
	fmt.Printf("+++ DOWNLOAD #4: EXPECTED 'COMPLETE: CONTENT', GOT '%s'\n", fourthDownload) // this... is not fine - it shows CANCELED
}
```
</details>

**Comment by ZekeLu** [comment]
Hmm, the workaround does not work actually. The target should belong to the same browser context too. I have added a commit (https://github.com/ZekeLu/chromedp/commit/a69be0ba1597d996cb97004c2c43c8f4059966fb) to create a target in a browser context several months ago. But I don't know any use case then, so it's not merged. I will try to add some tests and get it merged. After then, I will update the workaround.

**Comment by onsi** [comment]
thanks for digging into this so quickly @ZekeLu !  I'm happy to help whatever way I can (testing against a branch/patch/etc; opening an issue on chromium if need be - thought it sounds like that isn't the issue).

I suspected this had to do with the distinction between `page` and `browser` and saw the deprecation of `page.SetDownloadBehavior` but wasn't sure if `browser.SetDownloadBehavior` needed to be sent to the individual targets or to "the browser" and - if the browser - I wasn't sure how to do that.  I also noticed that there are always pairs of events generated (e.g.):
```
2023/02/23 22:32:38 <- {"method":"Page.downloadProgress","params":{"guid":"435d6fe9-5ab8-4dd4-b415-697520528569","totalBytes":0,"receivedBytes":0,"state":"inProgress"},"sessionId":"7912B8EB4AA7F8C7FB2C4C82B4D55CDE"}
2023/02/23 22:32:38 <- {"method":"Browser.downloadProgress","params":{"guid":"435d6fe9-5ab8-4dd4-b415-697520528569","totalBytes":0,"receivedBytes":7,"state":"inProgress"},"sessionId":"7912B8EB4AA7F8C7FB2C4C82B4D55CDE"}
```

and that `chromedp` as sending down the `Page.downloadProgress` events and I couldn't find a way to get the `Browser.downloadProgress` events; it appears thate `ListenBrowser` doesn't do it  because these events have a `sessionId` and `method` [so they get routed](https://github.com/chromedp/chromedp/blob/448735f388d4cca1619c64d748412b037687a0a3/browser.go#L254-L259) to the `ListenTarget` handler (I think)

In any event - am happy to try whatever workaround you end up rolling out when you have something.

**Comment by ZekeLu** [comment]
Here is the updated demo based on #1269:

```go
package main

import (
	"context"
	"fmt"
	"log"
	"net/http"
	"os"

	"github.com/chromedp/cdproto/browser"
	"github.com/chromedp/chromedp"
)

const DOWNLOAD_PAGE = `<!DOCTYPE html>
<html lang="en-US">
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width" />
    <title>Downloads Testpage</title>
</head>
<body>
    <button id="download">Download</button>
    <script>
        document.getElementById("download").addEventListener("click", () => {
            let anchor = document.createElement('a');
            anchor.setAttribute('href', 'data:text/plain;charset=utf-8,CONTENT');
            anchor.setAttribute('download', 'file.txt');
            anchor.click()
        })
    </script>
</body>
</html>`

func page(w http.ResponseWriter, req *http.Request) {
	fmt.Fprint(w, DOWNLOAD_PAGE)
}

func startServer() {
	http.HandleFunc("/page", page)
	go http.ListenAndServe(":8080", nil)
}

func main() {
	startServer()

	// first we start up a browser instance and grab its websocket url
	fmt.Print("+++ SPINNING UP THE BROWSER\n\n")
	opts := append(chromedp.DefaultExecAllocatorOptions[:],
		chromedp.Flag("remote-debugging-port", "9222"),
	)
	execAllocCtx, cancelAllocCtx := chromedp.NewExecAllocator(context.Background(), opts...)
	defer cancelAllocCtx()
	startBrowserCtx, cancelStartBrowserCtx := chromedp.NewContext(execAllocCtx)
	defer cancelStartBrowserCtx()
	if err := chromedp.Run(startBrowserCtx); err != nil {
		panic(err)
	}

	wsURL := "ws://127.0.0.1:9222"

	//now we attach to the browser twice
	allocCtxA, canceldAlloCtxA := chromedp.NewRemoteAllocator(context.Background(), wsURL)
	defer canceldAlloCtxA()
	allocCtxB, canceldAlloCtxB := chromedp.NewRemoteAllocator(context.Background(), wsURL)
	defer canceldAlloCtxB()

	//we create a new tab on A
	fmt.Print("+++ CREATING FIRST TAB\n\n")
	tabA, cancelA := chromedp.NewContext(allocCtxA, chromedp.WithDebugf(log.Printf), chromedp.WithNewBrowserContext())
	defer cancelA()

	//and configure it for downloads
	fmt.Print("+++ CONFIGURING FIRST TAB DOWNLOAD BEHAVIOR\n\n")
	if err := chromedp.Run(tabA,
		chromedp.ActionFunc(func(ctx context.Context) error {
			return browser.SetDownloadBehavior(browser.SetDownloadBehaviorBehaviorAllowAndName).
				WithDownloadPath("./").
				WithBrowserContextID(chromedp.FromContext(ctx).BrowserContextID).
				WithEventsEnabled(true).Do(ctx)
		}),
	); err != nil {
		panic(err)
	}

	downloads := make(chan string, 1)
	chromedp.ListenTarget(tabA, func(ev interface{}) {
		switch ev := ev.(type) {
		case *browser.EventDownloadWillBegin:
			fmt.Print("\n+++ DOWNLOAD BEGIN\n\n")
		case *browser.EventDownloadProgress:
			switch ev.State {
			case browser.DownloadProgressStateCanceled:
				fmt.Print("\n+++ DOWNLOAD CANCELED\n\n")
				downloads <- "CANCELED"
			case browser.DownloadProgressStateCompleted:
				fmt.Print("\n+++ DOWNLOAD COMPLETE\n\n")
				content, _ := os.ReadFile("./" + ev.GUID)
				downloads <- "COMPLETE: " + string(content)
			}
		}
	})

	//and we try downloading something
	fmt.Print("\n+++ LOADING PAGE AND PERFORMING FIRST DOWNLOAD\n\n")
	if err := chromedp.Run(tabA,
		chromedp.Navigate(`http://localhost:8080/page`),
		chromedp.Click(`#download`, chromedp.ByQuery),
	); err != nil {
		panic(err)
	}
	firstDownload := <-downloads

	//and we do it again, to prove that we can
	fmt.Print("+++ PERFORMING SECOND DOWNLOAD\n\n")
	if err := chromedp.Run(tabA, chromedp.Click(`#download`, chromedp.ByQuery)); err != nil {
		panic(err)
	}
	secondDownload := <-downloads

	//now we create a new tab in the separate remote allocator
	fmt.Print("+++ CREATING NEW TAB\n\n")
	tabB, cancelB := chromedp.NewContext(allocCtxB, chromedp.WithDebugf(log.Printf), chromedp.WithNewBrowserContext())

	//and we configure it for downloads
	fmt.Print("+++ CONFIGURING NEW TAB DOWNLOAD BEHAVIOR\n\n")
	if err := chromedp.Run(tabB,
		chromedp.ActionFunc(func(ctx context.Context) error {
			return browser.SetDownloadBehavior(browser.SetDownloadBehaviorBehaviorAllowAndName).
				WithDownloadPath("./").
				WithBrowserContextID(chromedp.FromContext(ctx).BrowserContextID).
				WithEventsEnabled(true).Do(ctx)
		}),
	); err != nil {
		panic(err)
	}

	// back on tabA, we try another download
	fmt.Print("\n+++ PERFORMING THIRD DOWNLOAD (ON ORIGINAL TAB)\n\n")
	if err := chromedp.Run(tabA, chromedp.Click(`#download`, chromedp.NodeVisible)); err != nil {
		panic(err)
	}
	thirdDownload := <-downloads

	// now we close tab B
	fmt.Print("+++ CLOSING NEW TAB\n\n")
	cancelB()

	// and try one last download on tab A
	fmt.Print("\n+++ PERFORMING FOURTH DOWNLOAD (ON ORIGINAL TAB)\n\n")
	chromedp.Run(tabA, chromedp.Click(`#download`, chromedp.NodeVisible))
	fourthDownload := <-downloads

	fmt.Printf("+++ DOWNLOAD #1: EXPECTED 'COMPLETE: CONTENT', GOT '%s'\n", firstDownload)  // this is fine
	fmt.Printf("+++ DOWNLOAD #2: EXPECTED 'COMPLETE: CONTENT', GOT '%s'\n", secondDownload) // this is fine
	fmt.Printf("+++ DOWNLOAD #3: EXPECTED 'COMPLETE: CONTENT', GOT '%s'\n", thirdDownload)  // this is fine
	fmt.Printf("+++ DOWNLOAD #4: EXPECTED 'COMPLETE: CONTENT', GOT '%s'\n", fourthDownload) // this is fine
}
```

**Comment by onsi** [comment]
hey it's looking good!  I had a few minutes so I hacked the PR changes in and am seeing promising behavior.  I'm going to stress-test things some more on my end but I'm happy to say this is likely fixed now :)

I think I just have one question at this point.  I have control over things such that when a user explicitly requests a new tab I can create in in a new browser context.

However, if a user action (e.g. clicking an anchor tag with `target="_blank"`) results in Chrome opening a new tab I _believe_ I lose that control.  Correct?  My assumption is that tab/target will open within the browser context of the tab/target that opened it - which means that when I attach to it and `SetDownloadBehavior` I will run into issues on the original tab when that tab is closed.  (I'll validate that this, in fact, continues to happen).

I, further, assume that there is no way to "move" that target to its own `BrowserContext`.  i.e.

`chromedp.NewContext(originalTarget.Context, chromedp.WithTargetID(newTarget.TargetID), chromedp.WithNewBrowserContext())`

won't work, will it?  (lol, sorry I should just try this instead of bothering you.... will do and report back)

**Comment by ZekeLu** [comment]
> * this issue does **not** occur if I create `tabA` and `tabB` using the `ExecAllocator`:
> 
> ```
> // this is fine:
> tabA, _ := chromedp.NewContext(execAllocCtx)
> tabB, _ := chromedp.NewContext(execAllocCtx)
> ```

I forgot to answer this question. This code does not have the issue because `tabA` and `tabB` are in difference browser instances. See the doc below (Line 90~92)

https://github.com/chromedp/chromedp/blob/448735f388d4cca1619c64d748412b037687a0a3/chromedp.go#L86-L100

---

> in your opinion can I share a code sample that uses chromedp? or do they prefer i speak purely in terms of the protocol?

I think it's better to show the issue using https://github.com/puppeteer/puppeteer. I created https://github.com/ZekeLu/cdp-download-events to report https://bugs.chromium.org/p/chromium/issues/detail?id=1204880 before.

---

> However, if a user action (e.g. clicking an anchor tag with `target="_blank"`) results in Chrome opening a new tab I _believe_ I lose that control. Correct? My assumption is that tab/target will open within the browser context of the tab/target that opened it

You're correct.

---

> which means that when I attach to it and `SetDownloadBehavior` I will run into issues on the original tab when that tab is closed.

This is a good question. Then I think we have to go another way around. See the demo below. It does not use 'BrowserContext` at all.

<details><summary><code>chromedp.ListenBrowser</code> demo</summary>

```go
package main

import (
	"context"
	"fmt"
	"log"
	"net/http"
	"os"

	"github.com/chromedp/cdproto/browser"
	"github.com/chromedp/cdproto/cdp"
	"github.com/chromedp/cdproto/target"
	"github.com/chromedp/chromedp"
)

const DOWNLOAD_PAGE = `<!DOCTYPE html>
<html lang="en-US">
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width" />
    <title>Downloads Testpage</title>
</head>
<body>
    <button id="download">Download</button>
    <a id="link" target="_blank" href="/page">New Page</button>
    <script>
        document.getElementById("download").addEventListener("click", () => {
            let anchor = document.createElement('a');
            anchor.setAttribute('href', 'data:text/plain;charset=utf-8,CONTENT');
            anchor.setAttribute('download', 'file.txt');
            anchor.click()
        })
    </script>
</body>
</html>`

func page(w http.ResponseWriter, req *http.Request) {
	fmt.Fprint(w, DOWNLOAD_PAGE)
}

func startServer() {
	http.HandleFunc("/page", page)
	go http.ListenAndServe(":8080", nil)
}

func main() {
	startServer()

	// browser
	browserCtx, cancel := chromedp.NewContext(context.Background(), chromedp.WithDebugf(log.Printf))
	defer cancel()
	if err := chromedp.Run(browserCtx,
		chromedp.ActionFunc(func(ctx context.Context) error {
			c := chromedp.FromContext(ctx)
			return browser.SetDownloadBehavior(browser.SetDownloadBehaviorBehaviorAllowAndName).
				WithDownloadPath("./").
				// use the Browser executor so that it does not pass "sessionId" to the command.
				WithEventsEnabled(true).Do(cdp.WithExecutor(ctx, c.Browser))
		}),
	); err != nil {
		panic(err)
	}

	downloads := make(chan string, 1)
	// Thanks to cdp.WithExecutor(ctx, c.Browser), now we can use ListenBrowser
	// to get the download events.
	chromedp.ListenBrowser(browserCtx, func(ev interface{}) {
		switch ev := ev.(type) {
		case *browser.EventDownloadWillBegin:
			fmt.Print("\n+++ DOWNLOAD BEGIN\n\n")
		case *browser.EventDownloadProgress:
			switch ev.State {
			case browser.DownloadProgressStateCanceled:
				fmt.Print("\n+++ DOWNLOAD CANCELED\n\n")
				downloads <- "CANCELED"
			case browser.DownloadProgressStateCompleted:
				fmt.Print("\n+++ DOWNLOAD COMPLETE\n\n")
				content, _ := os.ReadFile("./" + ev.GUID)
				downloads <- "COMPLETE: " + string(content)
			}
		}
	})

	// tab A
	tabA, cancel := chromedp.NewContext(browserCtx)
	defer cancel()

	if err := chromedp.Run(tabA,
		chromedp.Navigate(`http://localhost:8080/page`),
		chromedp.Click(`#download`, chromedp.ByQuery),
	); err != nil {
		panic(err)
	}
	firstDownload := <-downloads

	// <a target="_blank">
	ch := chromedp.WaitNewTarget(tabA, func(info *target.Info) bool {
		return info.URL == "http://localhost:8080/page"
	})
	if err := chromedp.Run(tabA,
		chromedp.Click(`#link`, chromedp.ByQuery),
	); err != nil {
		panic(err)
	}
	tabB, cancel := chromedp.NewContext(browserCtx, chromedp.WithTargetID(<-ch))
	defer cancel()
	if err := chromedp.Run(tabB,
		chromedp.Click(`#download`, chromedp.ByQuery),
	); err != nil {
		panic(err)
	}

	secondDownload := <-downloads

	fmt.Printf("+++ DOWNLOAD #1: EXPECTED 'COMPLETE: CONTENT', GOT '%s'\n", firstDownload)
	fmt.Printf("+++ DOWNLOAD #3: EXPECTED 'COMPLETE: CONTENT', GOT '%s'\n", secondDownload)
}
```
</details>

---

> I, further, assume that there is no way to "move" that target to its own `BrowserContext`. i.e.
> 
> `chromedp.NewContext(originalTarget.Context, chromedp.WithTargetID(newTarget.TargetID), chromedp.WithNewBrowserContext())`
> 
> won't work, will it?

It won't work. See the test `WithTargetID wins WithNewBrowserContext` and `WithTargetID wins WithExistingBrowserContext` in the PR. As far as I know, there is no way to "move" a target to another `BrowserContext`.

**Comment by onsi** [comment]
Thanks for all the help @ZekeLu  - I think with the PR you've put together and some careful accounting on my part I'm in a good place now.  Do you have a sense for when it will be merged and a release cut?

As I mentioned earlier, I'm working on a companion framework for Ginkgo and Gomega that is built on top of chromedp.  My goal is to allow for high-performance (and low-flakiness) browser integration testing by leveraging Ginkgo's support for test parallelization.  The current architecture looks like this:

- a single chrome browser is allocated
- each test process then attaches to this browser and creates a single tab
- this tab is reused between tests (creating a new tab in between tests, while cleaner, does introduce a performance hit)
- the user interacts with this tab in their tests, potentially spawning new tabs (either explicitly or implicitly if a javascript action results in a tab being created)

This architecture was working well until I hit this snag.  When running in parallel a tab closed on one process would clear out the `SetDownloadBehavior` for a tab in a different process.  Thankfully this is now resolved - since each explicitly created tab is created in its own `BrowserContext` I no longer have this issue.  In the case where the tab is spawned I can track that in-process and (a) prevent the tab from closing while a download with the same `BrowserContext` is taking place and (b) reset `SetDownloadBehavior` for all other tabs in that `BrowserContext` when tab does eventually close.

in this way I try to manage some of chrome's complexity for the user so they don't have to deal with flaky tests.  As another example, the framework I'm working on tracks the number of concurrent downloads and ensures that no more than 10 could ever occur wtihin a second.  This prevents users from bumping into #791 

So far the results are promising.  I have 153 tests that test the framework.  When I run them in series the test suite takes 8.5 seconds.  When I run them in parallel on my M1 Max machine they run in 1.8 seconds (!!)

Chromedp is great!  It's been a pleasure building on top of it :)

**Comment by ZekeLu** [comment]
> I'm working on a companion framework for Ginkgo and Gomega that is built on top of chromedp.

Good job! :tada: :tada: :tada: 

> Do you have a sense for when it will be merged and a release cut?

I will merge the PR and tag a release this weekend if everything goes well.

**Comment by onsi** [comment]
> I will merge the PR and tag a release this weekend if everything goes well.

perfect - thanks!!

**Comment by onsi** [comment]
hey - I think I've bumped into something new that is occurring with the new PR code.  It's not strictly download related.  I've updated the [reproducer](https://github.com/onsi/download-reproducer).   Here's what I'm seeing

### Background

In the thing I'm building I allow the user to list all open tabs and I use `chromedp.Targets` to fetch targets.  I do something like this to keep track of things:

```
targets, err := chromedp.Targets(rootTab)
for _, target := range targets {
	tab, ok := myTabs[target.TargetID]
	if ok {
		continue
	}
	// this may be a new tab we've never seen before - is it ours?
	opener := myTabs[target.OpenerID]
	if opener != nil {
		myTabs[target.TargetID] = chromedp.NewContext(opener.Context, chromedp.WithTargetID(target.TargetID))
	} 
}
```

this works well - the new tab context is created in the same `BrowserContextID` as the opener and I am able to keep things organized so that I can help the user answer questions like "show me all the tabs" and "show me the tabs that were opened by _this_ tab."  More importantly, all my download management stuff works since the `BrowserContextID`s are aligned.

But - I'm seeing this error message:

```
ERROR: WithExistingBrowserContext is ignored when WithTargetID is specified.
```

So I did some investigating...

### Reproducer

It looks like you _can_ actually assign a BrowserContextID when attaching to the target.  The new reproducer does the following:

1. Spin up a browser then connect to it via a remote allocator
2. Create and spin up two new contexts (`firstTab` and `secondTab`) -in the reproducer `firstTab` is created from the `allocCtx` and `secondTab` is created from the `firstTab` context - but that doesn't seem to matter.
3. On the `firstTab` navigate to a page with a `target="_blank"` link and click on the link
4. Wait a bit for the tab to open
5. List the tabs and find the one whose opener is `firstTab`.
6. Create a context using `chromedp.NewContext(firstTab, chromedp.WithTargetID(target.TargetID))`
7.  We see the error message, but also see that this new context has a BrowserContextID that is the same as `firstTab`
8. Close the new tab and click on the link in `firstTab` again
9. Find this _new_ tab, but this time create a context using `chromedp.NewContext(**secondTab**, chromedp.WithTargetID(target.TargetID))`
10. we see the same error message... but, surprisingly, the tab now has `secondTab`'s `BrowserContextID`!

Here's the output:

```
First Tab -  | ID:F12DB12FCE661BA7544E5F28A7F9D976 | Browser ID: A621FD041A5968EB30BC88E82F65B3CB
Second Tab -  | ID:676E8C93CA29334F864637D4962F6E9D | Browser ID: CD388A9B5B175B65123A207F1D94A371
2023/02/28 12:49:45 ERROR: WithExistingBrowserContext is ignored when WithTargetID is specified.
New Tab (context based off of firstTab) - Page B Testpage | ID:B3BF1F2AC672728B2924CE726C1E6F0C | Browser ID: A621FD041A5968EB30BC88E82F65B3CB
2023/02/28 12:49:46 ERROR: WithExistingBrowserContext is ignored when WithTargetID is specified.
New Tab (context based off of secondTab) - Page B Testpage | ID:1B3087819ED9FE6773CCBFB62D3AD0EB | Browser ID: CD388A9B5B175B65123A207F1D94A371
```

And, it seems that `BrowserContextID` is "real."  I haven't included this in the reproducer, but I find that `SetDownloadBehavior` is only undone for the `BrowserContextID` associated with the tab.

### The Issue

I know that's a lot of detail.  The `BrowserContextID` stuff is interesting as it seems to illustrate that the id can be chosen when creating the context.  The real issue is that error message: 

`2023/02/28 12:49:45 ERROR: WithExistingBrowserContext is ignored when WithTargetID is specified.`

it pops up when I attach to new tabs even though I'm not specifying `WithExistingBrowserContext` - I'm just creating the context using a parent context that happens to have a `BrowserContextID`.  It does seem, though, that the `BrowserContextID` is **not** ignored.

**Comment by ZekeLu** [comment]
Good catch! This is a bug in my implementation. I haven't thought of such a valid case that the context is inherited while the `BrowserContext` is not. I will try to fix it and update here once it's fixed.

Thank you!

**Comment by onsi** [comment]
Thanks!

**Comment by ZekeLu** [comment]
I have just updated the PR. There are two changes:

1. if the usages of the `ContextOption` is incorrect, it panics instead of writing error messages to the logger. This makes the new options align with `WithBrowserOption`.
2. it won't inherit the value of `BrowserContextID` from its parent context when the context is used to attach to an existing target. **And the value is empty in this case**.

The doc comment for `BrowserContextID` is updated to:

```go
// BrowserContextID is set up by WithExistingBrowserContext.
//
// Otherwise, BrowserContextID holds a non-empty value in the following cases:
//
// 1. if the context is created with the WithNewBrowserContext option, a new
// BrowserContext is created on its first run, and BrowserContextID holds
// the id of that new BrowserContext;
//
// 2. if the context is not created with the WithTargetID option, and its
// parent context has a non-empty BrowserContextID, this context's
// BrowserContextID is copied from the parent context.
BrowserContextID cdp.BrowserContextID
```

So we can not get the `BrowserContextID` from a context when it is used to attach to an existing tab. If we do want the value, we have to issue the `target.GetTargetInfo` to get it:

```go
func getBrowserContextID(ctx context.Context) (cdp.BrowserContextID, error) {
	var id cdp.BrowserContextID
	err := chromedp.Run(ctx,
		chromedp.ActionFunc(func(ctx context.Context) error {
			info, err := target.GetTargetInfo().Do(ctx)
			id = info.BrowserContextID
			return err
		}),
	)
	return id, err
}
```

**Comment by onsi** [comment]
hey @ZekeLu - thanks!  I just pulled this in and it took me a minute to figure out the changes but I believe I have things working now.  Chrome does seem to put spawned tabs in the correct `BrowserContextID` and I am able to retrieve that id with the `GetTargetInfo()` approach.  Most importantly, I no longer see the error messages getting printed and the target's `BrowserContextID` is not assumed to be inherited from the parent context.

The only new behavior that I noticed is that I now need to create and initialize a dummy empty tab for my `RemoteAllocator` before I can create a  tab with `chromedp.WithNewBrowserContext()`.  If I don't `chromedp` panics with 

```
panic: WithNewBrowserContext can not be used before Browser is initialized
```

this particular piece did seem to work fine before.  It's no problem on my end to make the empty tab but I'm just not sure that it's necessary given it seemed (?)  to be working before.

**Comment by ZekeLu** [comment]
Sorry, it's a stupid bug in my code. I just updated the PR. A regression test is pending. Please try again. Thank you!

**Comment by onsi** [comment]
looks good now!

**Comment by ZekeLu** [comment]
Regression tests have been added. Thank you!

**Comment by onsi** [comment]
🎉

**Comment by onsi** [comment]
hey @ZekeLu  - in case you're interested, I just open-sourced the thing I've been working on that led to this issue.  It's a library called Biloba, built on top of chromedp, that integrates with GInkgo and Gomega.  I know Ginkgo/Gomega aren't for everybody - so no expectations at all that you'd find this interesting or useful.  Just wanted to share so you were aware what's happening in the ecosystem.

Thanks for all your hard work!

[The Biloba Repo](https://github.com/onsi/biloba)
[The Biloba Docs](https://onsi.github.io/biloba/)
